### PR TITLE
Squashed third implementation

### PR DESCRIPTION
**Pull request by purple52**

### CODE REVIEW

File: src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            initialiseForTupleArrays(
+                    schema.getItemSchemas(),
+                    schema.permitsAdditionalItems(),
+                    Optional.ofNullable(schema.getSchemaOfAdditionalItems()));
+        } else if (schema.getItemSchemas() == null && schema.getAllItemSchema() == null) {
+            initialiseForUnspecifiedArrays();
+        } else {
+            throw new JsonGenerationException("Inconsistent array schema");
+        }
+    }
```
Comment (allanmckenzie) [inline src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java:56]: This is a lot of logic to put in a constructor. Would an init() method be better?

Comment (allanmckenzie) [reply src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java:56]: Is this fully tested? as in every path and every boolean condition? It's rather hard to see from this review code

**Review by allanmckenzie — APPROVED** [review]

File: src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+     * @param index the index of the array item to generate
+     * @return a value that matches the schema for that index
+     */
+    private JsonValue value(final int index) {
+        if (index >= maxItems) {
+            throw new JsonGenerationException("Tried to generate a value for an array index greater than the max length");
+        } else if (itemGenerators != null && index < itemGenerators.size()) {
+            return itemGenerators.get(index).next();
+        } else {
+            return itemGenerator.next();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java:117]: Can itemGenerator ever be null? Why have a single itemGenerator and a List of itemGenerators?

File: src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import javax.json.JsonValue;
+
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.Schema;
+
+public class JsonCombinedSchemaGenerator extends Generator<JsonValue> {
+
+    private Generator<? extends JsonValue> generator;
+
+    public JsonCombinedSchemaGenerator(CombinedSchema combinedSchema) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java]: final?

File: src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    private void getAnyOf(CombinedSchema combinedSchema) {
+        final Generator<Schema> schemaGenerator =
+                new ValueGenerator(combinedSchema.getSubschemas());
+        generator = generatorFor(schemaGenerator.next());
+    }
+
+    @Override
+    public JsonValue next() {
+        return generator.next();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java]: can generator be null?

File: src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        if(combinedSchema.getCriterion() == ONE_CRITERION ) {
+            throw new JsonGenerationException("oneOf combined schema type is not supported");
+        }
+
+        if(combinedSchema.getCriterion() == ALL_CRITERION ) {
+            throw new JsonGenerationException("allOf combined schema type is not supported");
+        }
+
+        if(combinedSchema.getCriterion() == ANY_CRITERION ) {
+            getAnyOf(combinedSchema);
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java]: The method name does not seem to describe what is occurring here

File: src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        }
+    }
+
+    private void getAnyOf(CombinedSchema combinedSchema) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonCombinedSchemaGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/JsonNumberGenerator.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        if (requiresInteger) {
+            numberGenerator = integerGenerator(numberSchema);
+        } else {
+            numberGenerator = bigDecimalGenerator(numberSchema);
+        }
+    }
+
+    private Generator<Integer> integerGenerator(final NumberSchema numberSchema) {
+        IntegerGenerator.Builder builder = IntegerGenerator.builder();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonNumberGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/JsonNumberGenerator.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            builder.exclusiveMaximum(true);
+        }
+        if (numberSchema.isExclusiveMinimum()){
+            builder.exclusiveMinimum(true);
+        }
+        return builder.build();
+    }
+
+    private Generator<BigDecimal> bigDecimalGenerator(final NumberSchema numberSchema) {
+        BigDecimalGenerator.Builder builder = BigDecimalGenerator.builder();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonNumberGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/JsonObjectGenerator.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    private Map<String, Generator<? extends JsonValue>> generators;
+    private Set<String> properties;
+    private Set<String> requiredProperties;
+
+    public JsonObjectGenerator(final ObjectSchema schema) {
+        properties = schema.getPropertySchemas().keySet();
+        requiredProperties = new HashSet<>(schema.getRequiredProperties());
+
+        generators = new HashMap<>();
+        Map<String, Schema> propertySchemas = schema.getPropertySchemas();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonObjectGenerator.java]: Multiple calls to schema.getPropertySchemas(), need only call once.

File: src/main/java/uk/gov/justice/json/generator/JsonReferenceGenerator.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+
+public class JsonReferenceGenerator extends Generator<JsonValue> {
+
+    private final Schema referrencedSchema;
+
+
+    public JsonReferenceGenerator(ReferenceSchema referenceSchema) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonReferenceGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/ValueGenerator.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Random;
+
+public class ValueGenerator extends Generator {
+
+    private final Collection possibleValues;
+    private final int value;
+
+    public ValueGenerator(Collection possibleValues) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/ValueGenerator.java]: possibleValues can be final

File: src/main/java/uk/gov/justice/json/generator/ValueGenerator.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        this.possibleValues = possibleValues;
+        final int size = this.possibleValues.size();
+        Random random = new Random();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/ValueGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/jsonvalue/EnumJsonValueGenerator.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.everit.json.schema.EnumSchema;
+import org.everit.json.schema.Schema;
+import org.json.JSONObject;
+
+public class EnumJsonValueGenerator extends Generator<JsonValue> {
+
+    private final Set<Object> possibleValues;
+
+    public EnumJsonValueGenerator(EnumSchema schema) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/jsonvalue/EnumJsonValueGenerator.java]: schema can be final

File: src/main/java/uk/gov/justice/json/generator/jsonvalue/EnumJsonValueGenerator.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+            if (value.equals(Boolean.TRUE)) {
+                return JsonValue.TRUE;
+            }
+            return JsonValue.FALSE;
+        }
+         throw new JsonGenerationException(format("Unsupported Type: %s",value ));
+
+    }
+
+    private Object randomlyGet(Set<Object> possibleValues) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/jsonvalue/EnumJsonValueGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    private static final long DEFAULT_MIN = -1000000000L;
+    private static final long DEFAULT_MAX = 1000000000L;
+    private Optional<BigDecimal> minimum = Optional.of(BigDecimal.valueOf(DEFAULT_MIN));
+    private Optional<BigDecimal> maximum = Optional.of(BigDecimal.valueOf(DEFAULT_MAX));
+    private Optional<BigDecimal> multipleOf;
+    private boolean exclusiveMinimum;
+    private boolean exclusiveMaximum;
+    private int scale = 1;
+
+    private BigDecimalGenerator(BigDecimalGenerator.Builder builder) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java]: builder can be final

File: src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+        public Builder() {
+
+        }
+
+        public BigDecimalGenerator build() {
+            return new BigDecimalGenerator(this);
+        }
+
+        public BigDecimalGenerator.Builder exclusiveMaximum(boolean exclusiveMaximum) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java]: can be final and in other methods below

File: src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    private static final Integer DEFAULT_MIN = Integer.MIN_VALUE;
+    private static final Integer DEFAULT_MAX = Integer.MAX_VALUE;
+    private final Random random = new Random();
+    private boolean exclusiveMinimum;
+    private boolean exclusiveMaximum;
+    private Optional<Integer> minimum;
+    private Optional<Integer> maximum;
+    private Optional<Integer> multipleOf;
+
+    private IntegerGenerator(IntegerGenerator.Builder builder) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java]: final

File: src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+
+    public static IntegerGenerator.Builder builder() {
+        return new IntegerGenerator.Builder();
+    }
+
+    public Integer next() {
+        int minimum = this.minimum.get();
+        int maximum = this.maximum.get();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java:47]: these could be final

File: src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        private boolean exclusiveMaximum = false;
+
+        public Builder() {
+        }
+
+        public IntegerGenerator build() {
+            return new IntegerGenerator(this);
+        }
+
+        public IntegerGenerator.Builder exclusiveMaximum(boolean exclusiveMaximum) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/number/IntegerGenerator.java]: can be final and methods below

File: src/main/java/uk/gov/justice/json/generator/value/string/RegexGenerator.java [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.regex.Pattern;
+
+import com.mifmif.common.regex.Generex;
+
+public class RegexGenerator extends Generator<String> {
+
+    private final Pattern pattern;
+
+    public RegexGenerator(Pattern pattern) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/string/RegexGenerator.java]: pattern can be final

File: src/main/java/uk/gov/justice/json/generator/value/string/SimpleStringGenerator.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+import uk.gov.justice.services.test.utils.core.random.StringGenerator;
+
+import org.apache.commons.lang3.RandomStringUtils;
+
+public class SimpleStringGenerator extends Generator<String> {
+
+    private int minLength =10;
+    private final int maxLength;
+
+    public SimpleStringGenerator(int minLength, int maxLength) {
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/value/string/SimpleStringGenerator.java]: final

File: src/test/java/uk/gov/justice/json/MatcherBasedTypeCheck.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        return this;
+    }
+
+    public void verify() {
+        verify(Times.times(1));
+    }
+
+    public void verify(Times times) {
+        for (int i = 0; i < times.getNum(); i++) {
+            T next = generator.next();
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/MatcherBasedTypeCheck.java]: final

File: src/test/java/uk/gov/justice/json/MatcherBasedTypeCheck.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+            } else {
+                assertThat(format("failed on attempt %s for value %s", i + 1, next), next, matcher);
+            }
+        }
+    }
+
+    public static class Times {
+        private final int num;
+
+        private Times(int num) {
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/MatcherBasedTypeCheck.java]: final and below

File: src/test/java/uk/gov/justice/json/generator/EnumJsonValueGeneratorTest.java [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+public class EnumJsonValueGeneratorTest {
+
+    private static final int NUMBER_OF_TIMES = 1000;
+
+    @Test
+    public void shouldGenerateStringValue() throws IOException, URISyntaxException {
+
+        final Schema schema = JsonSchemaLoader.loadSchema("src/test/resources/schemas/enums/simple.enum.json-schema.json");
+
+        EnumJsonValueGenerator jsonEnumGenerator = new EnumJsonValueGenerator((EnumSchema) schema);
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/EnumJsonValueGeneratorTest.java]: final

File: src/test/java/uk/gov/justice/json/generator/EnumJsonValueGeneratorTest.java [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+                .verify(times(NUMBER_OF_TIMES));
+
+    }
+
+    @Test
+    public void shouldGenerateNullValue() {
+
+        final Schema schema = JsonSchemaLoader.loadSchema("src/test/resources/schemas/enums/null.enum.json-schema.json");
+
+        EnumJsonValueGenerator jsonEnumGenerator = new EnumJsonValueGenerator((EnumSchema) schema);
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/EnumJsonValueGeneratorTest.java]: final and other tests below

File: src/test/java/uk/gov/justice/json/generator/JsonStringGeneratorTest.java [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+
+        final StringSchema stringSchema = new StringSchema().builder().formatValidator(new IPV6Validator()).build();
+        final JsonStringGenerator jsonStringGenerator = new JsonStringGenerator(stringSchema);
+
+        typeCheck(jsonStringGenerator, jsonString -> new IPV6Validator().validate(((JsonString)jsonString).getString())
+                .equals(Optional.empty()))
+                .verify(times(NUMBER_OF_TIMES));;
+    }
+
+    private Optional<String> validate(FormatValidator validator,JsonValue jsonValue){
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/JsonStringGeneratorTest.java]: final

File: src/test/java/uk/gov/justice/json/generator/matcher/JsonValueSchemaMatcher.java [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            @Override
+            protected void describeMismatchSafely(final JsonValue jsonValue, final Description mismatchDescription) {
+                mismatchDescription.appendText("validation failed with message ").appendValue(exception.getMessage());
+            }
+        };
+    }
+
+    private static Boolean validateJsonObject(Schema schema, JsonValue jsonValue) {
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/matcher/JsonValueSchemaMatcher.java]: final

File: src/test/java/uk/gov/justice/json/generator/multivalue/ArrayGeneratorTest.java [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+    @Test
+    public void shouldGenerateAJsonArrayValue() throws Exception {
+
+        final String intValue = "23";
+        final String stringValue = "\"Fred Bloggs\"";
+        final String objectValue = "{\"property\": true}";
+//
+//        final JsonArrayGenerator arrayValueGenerator = new JsonArrayGenerator(asList(intValue, stringGenerator, objectGenerator));
+//
+//        assertThat(arrayValueGenerator.nextValue(), is("[23,\"Fred Bloggs\",{\"property\": true}]"));
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/multivalue/ArrayGeneratorTest.java]: nothing being tested here, commented out assertion

File: src/test/java/uk/gov/justice/json/generator/value/EmailGeneratorTest.java [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+import org.apache.commons.validator.routines.EmailValidator;
+import org.junit.Test;
+
+public class EmailGeneratorTest {
+
+    @Test
+    public void shouldGenerateARandomEmail() throws Exception {
+        final EmailGenerator emailGenerator = new EmailGenerator();
+        final String jsonEmailValue = emailGenerator.next();
+        System.out.println(jsonEmailValue);
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/value/EmailGeneratorTest.java]: System.out can be removed

File: src/test/java/uk/gov/justice/json/generator/value/StringValueGeneratorTest.java [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+    @Test
+    public void shouldGenerateARandomString() throws Exception {
+        SimpleStringGenerator stringValueGenerator = new SimpleStringGenerator(4,12);
+        final String value = stringValueGenerator.next();
+        assertThat(value, isA(String.class));
+    }
+
+    @Test
+    public void shouldGenerateARandomStringWithMinimumLength() throws Exception {
+        SimpleStringGenerator stringValueGenerator = new SimpleStringGenerator(4,12);
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/value/StringValueGeneratorTest.java]: final

File: src/test/java/uk/gov/justice/json/generator/value/StringValueGeneratorTest.java [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+import uk.gov.justice.json.generator.value.string.SimpleStringGenerator;
+
+import org.junit.Test;
+
+public class StringValueGeneratorTest {
+
+    @Test
+    public void shouldGenerateARandomString() throws Exception {
+        SimpleStringGenerator stringValueGenerator = new SimpleStringGenerator(4,12);
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/value/StringValueGeneratorTest.java]: final

File: src/test/java/uk/gov/justice/json/generator/value/UriGeneratorTest.java [hunk 32]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.junit.Test;
+
+public class UriGeneratorTest {
+
+    @Test
+    public void shouldGenerateUri(){
+        final String uriAddress = new UriGenerator().next();
+//        final RegexValidator regexValidator = new RegexValidator(UriGenerator.EMAIL_PATTERN_REGEX, true);
+//        assertThat(true, is(regexValidator.isValid(uriAddress)));
```
Comment (mapingo) [inline src/test/java/uk/gov/justice/json/generator/value/UriGeneratorTest.java]: nothing tested here

**Review by mapingo — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java [hunk 33]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    private void initialiseForUnspecifiedArrays() {
+        itemGenerator = new JsonValueGenerator(EmptySchema.INSTANCE);
+    }
+
+    @Override
+    public JsonArray next() {
+        final int numItems = ThreadLocalRandom.current().nextInt(minItems, maxItems + 1);
+        JsonArrayBuilder builder = createArrayBuilder();
```
Comment (mapingo) [inline src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java]: could be final

File: src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java (hunk 1 above)
Comment (purple52) [reply src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java:56]: It isn't really logic - it's extracting from the schema and setting up the generator. We have spiked and compared with static builder methods (see https://github.com/CJSCommonPlatform/json-generator/tree/factory-methods). The code was slightly nicer, but it doesn't buy us enough right now to rewrite everything. We might want to refactor gradually after this is merged.

File: src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java (hunk 2 above)
Comment (purple52) [reply src/main/java/uk/gov/justice/json/generator/JsonArrayGenerator.java:117]: This weirdness is due to how JSON schemas work. We should refactor at some point to make this cleaner.

File: src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java (hunk 16 above)
Comment (purple52) [reply src/main/java/uk/gov/justice/json/generator/value/number/BigDecimalGenerator.java]: Not going to fix these much. BigDecimalGenerator and IntegerGenerator need rewriting.

**Review by mapingo — APPROVED** [review]

**Pull request merged**